perceval/client.py

```diff
@@ -0,0 +1,236 @@
+# the Free Software Foundation; either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA.
+#
+# Authors:
+#     Valerio Cosentino <valcos@bitergia.com>
+#
+
+import logging
+import time
+
+import requests
+from requests.adapters import HTTPAdapter
+from urllib3.util import Retry
+
+from .errors import RateLimitError
+
+logger = logging.getLogger(__name__)
+
+
+class HttpClient:
+    """Abstract class for HTTP clients.
+
+    Base class to query data sources taking care of retrying requests in case connection issues. If the data source does
+    not send back a response after retrying a request, a RetryError exception is thrown.
+
+    Sub-classes can use the methods fetch and _fetch to obtain data from the data source.
+
+    To track which version of the client was used during the fetching
+    process, this class provides a `version` attribute that each client
+    may override.
+
+    :param base_url: base URL of the data source
+    :param max_retries: number of max retries to a data source before raising a RetryError exception
+    :param max_retries_on_connect: max number of retries on connection error
+    :param max_retries_on_read: max number of retries on read errors
+    :param max_retries_on_redirect: max number of retries on redirects
+    :param max_retries_on_status: max number of retries on selected error status codes
+    :param status_forcelist: list of status codes where the retry attempts happen
+    :param method_whitelist: list of uppercased HTTP method verbs where the retry happen
+    :param raise_on_redirect: try retries on redirect status codes
+    :param raise_on_status: try retries on selected status codes
+    :param respect_retry_after_header: try retries for status code with retry-after headers
+    :param default_sleep_time: default time to sleep in case of connection problems
+    :param headers: list of session headers
+    """
+    version = '0.1'
+
+    DEFAULT_SLEEP_TIME = 1
+
+    MAX_RETRIES = 5
+    MAX_RETRIES_ON_CONNECT = 5
+    MAX_RETRIES_ON_READ = 5
+    MAX_RETRIES_ON_REDIRECT = 5
+    MAX_RETRIES_ON_STATUS = 5
+
+    DEFAULT_METHOD_WHITELIST = False
+    DEFAULT_RAISE_ON_REDIRECT = True
+    DEFAULT_RAISE_ON_STATUS = True
+    DEFAULT_RESPECT_RETRY_AFTER_HEADER = True
+
+    DEFAULT_RETRY_AFTER_STATUS_CODES = [413, 429, 503]
+    DEFAULT_STATUS_FORCE_LIST = [408, 423, 504]
+
+    GET = "GET"
+    POST = "POST"
+
+    def __init__(self, base_url, max_retries=MAX_RETRIES, max_retries_on_connect=MAX_RETRIES_ON_CONNECT,
```

Comment:
Do we really need all these configuration parameters? I think we should stick to only those we will use for sure. If we need more, we will add them in future versions.